### PR TITLE
[lit-next] Add Attribute, Property, and BooleanAttribute parts

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: packages/lit-html/src/lib/lit-html.ts [hunk 1]
```diff
@@ -560,6 +580,119 @@ export class NodePart {
...
+   * @param value the raw input value to normalize
+   * @param _i the index in the values array this value was read from
+   */
+  __getValue(value: unknown, _i: number) {
+    // TODO (justinfagnani): invoke directives here, which will need
+    // _i to revive directive state
+    return value == null ? '' : value;
+  }
+
+  __setValue(value: unknown | Array<unknown>, from?: number) {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/lit-html/src/lib/lit-html.ts:623]: The implementation is so much nicer! Appears to solve most of the weird edge cases that came up when I was doing the directives tests for SSR:

- [x] Any value that resolve to `nothing` in an `AttributePart` causes the entire attribute to be removed (logical OR)
- [x] A value that resolves to `noChange` uses its previous value
- [x] A value that initially resolves to `noChange` won't try to render `undefined`
- [x] Values start out as `nothing`, avoiding mutations during first render after hydration when initially rendering `nothing`

These were also weird things that came up in those tests don't look like they'll be solved yet:
- [ ] Rendering `nothing` to a `PropertyPart` sets the `nothing` object value to the property; should we coerce `nothing` to `undefined` for properties?
- [ ] Rendering `nothing` to a `BooleanAttributePart` results in `true`; `removeAttribute` is probably the right thing

File: packages/lit-html/src/lib/lit-html.ts [hunk 2]
```diff
@@ -560,6 +580,119 @@ export class NodePart {
...
+    name: string,
+    strings: ReadonlyArray<string>,
+    _options?: RenderOptions
+  ) {
+    this.__element = element;
+    this.name = name;
+    if (strings.length > 2 || strings[0] !== '' || strings[1] !== '') {
+      this.__value = new Array(strings.length - 1).fill(nothing);
+      this.__strings = strings;
+    }
```
Comment (kevinpschaaf) [inline packages/lit-html/src/lib/lit-html.ts:606]:
```suggestion
    } else {
      this.__value = nothing;
    }
```

Seems like we should initialize single-value bindings' value to `nothing` so that an initial render of `nothing` dirty checks to no change?

File: packages/lit-html/src/test/lit-html_test.ts [hunk 3]
```diff
@@ -343,4 +471,315 @@ suite('lit-html', () => {
...
+        html`<a href="${'foo'}" class="bar"><div id=${'a'}></div></a>`,
+        container
+      );
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        `<a class="bar" href="foo"><div id="a"></div></a>`
+      );
+    });
+
+    test('renders to an attribute without quotes', () => {
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/lit-html_test.ts:640]: How about a test for multiple unquoted attribute bindings?

File: packages/lit-html/src/test/lit-html_test.ts [hunk 4]
```diff
@@ -343,4 +471,315 @@ suite('lit-html', () => {
...
+
+    test('renders undefined in attributes', () => {
+      render(html`<div attribute="${undefined}"></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div attribute=""></div>'
+      );
+    });
+
+    test('nothing sentinel removes an attribute', () => {
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/lit-html_test.ts:718]: Suggest adding tests for updates (not just initial render), specifically:
* Update to `nothing` removes attribute after having rendered one
* Update to `noChange` after having rendered something results in no DOM mutations

File: packages/lit-html/src/test/lit-html_test.ts [hunk 5]
```diff
@@ -343,4 +471,315 @@ suite('lit-html', () => {
...
+    });
+
+    test('removes attributes for true values', () => {
+      render(html`<div ?foo=${false}></div>`, container);
+      assert.equal(stripExpressionComments(container.innerHTML), '<div></div>');
+    });
+  });
+
+  suite('properties', () => {
+    test('sets properties', () => {
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/lit-html_test.ts:778]: Suggest adding test for `nothing` & `noChange` (both initial render and update)

File: packages/lit-html/src/test/lit-html_test.ts [hunk 6]
```diff
@@ -343,4 +471,315 @@ suite('lit-html', () => {
...
+  suite('boolean attributes', () => {
+    test('adds attributes for true values', () => {
+      render(html`<div ?foo=${true}></div>`, container);
+      assert.equal(
+        stripExpressionComments(container.innerHTML),
+        '<div foo=""></div>'
+      );
+    });
+
+    test('removes attributes for true values', () => {
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/lit-html_test.ts:771]: Suggest adding test for `nothing` & `noChange` (both initial render and update)

**Comment by justinfagnani** [comment]
PTAL

File: packages/lit-html/src/test/lit-html_test.ts [hunk 7]
```diff
@@ -699,6 +701,18 @@ suite('lit-html', () => {
       ]);
     });
 
+    test('renders to multiple attribute expressions without quotes', () => {
+      render(
+        html`<div foo=${'Foo'} bar=${'Bar'} baz=${'Baz'}></div>`,
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/lit-html/src/test/lit-html_test.ts:706]: That doesn't work in 1.x. Here is `<div foo=${'Foo'}bar=${'Bar'}></div>` renders `<div foo="Foobar=Bar">`, which is arguably correct. The other arguably correct option is `<div foo="Foo" bar="Bar">`.

What's interesting is that `<div foo=${'Foo'}${'Bar'}></div>` now renders to `<div foo="FooBar"></div>`.

Do we want to officially support that?

[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/lit-html/src/test/lit-html_test.ts:706]: Yeah, I think we shouldn't sort of infer "whitespace" semantics of adding an expression (since the user can and almost always will add it themselves if that's what they intend), and so the current behavior seems both correct and good to me for both of those cases.

[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/lit-html/src/test/lit-html_test.ts:706]: Ok. It locks us into the pretty sophisticated scanner, but I guess we're committed to that. Added a test.

File: packages/lit-html/src/test/lit-html_test.ts [hunk 8]
```diff
@@ -775,12 +803,14 @@ suite('lit-html', () => {
         '<div foo="A"></div>',
         'A'
       );
+      const observer = new MutationObserver(() => {});
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/lit-html/src/test/lit-html_test.ts:806]: eek 😬

**Comment by justinfagnani** [comment]
PTAL

File: packages/lit-html/src/test/lit-html_test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/lit-html/src/test/lit-html_test.ts:817]: Since the attribute is bound to div in the container, not the container yourself, I think you need to observe `container.firstChild`, or else `{attributes: true, subtree: true}`? I guess the latter is probably better so you catch anything.
